compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -5019,6 +5021,86 @@ public void visit(BLangForeach foreach) {
         }
     }
 
+    private boolean canEliminateIterator(BLangForeach loop) {
+        BLangExpression collection = loop.collection;
+        return switch (collection.getKind()) {
+            case BINARY_EXPR -> true;
+            case SIMPLE_VARIABLE_REF -> collection.expectedType.getKind() == TypeKind.ARRAY;
```

Comment:
why can't we extend this to support list-constructors and other expressions where the expected type is an array?

Reply:
Fixed